refactor(newsletter-sign-up): refactor by using the generic components

nathanbowang opened the pull request:
Refactor by using the generic Button, Form, Input, Section, SuccessConfirmation component.
Add 'Add Another Email' button to the NewsletterSignUp component.

nathanbowang committed:
refactor(newsletter-sign-up): refactor by using the generic components

Refactor by using the generic Button, Form, Input, Section, SuccessConfirmation component.
Add 'Add Another Email' button to the NewsletterSignUp component.

reubenberghan commented on components/Main/Button/index.tsx:
```diff
@@ -6,9 +6,9 @@ export type ButtonProps = {
 }
 
-export default function Button({ children, onClick }: ButtonProps): JSX.Element {
+export default function Button({ children, onClick: handleClick }: ButtonProps): JSX.Element {
```
For me the `handle` prefix is better used for the handler definition. Say if you were defining the handler in the body of the component. Here we are just passing the function down `onClick` is fine to me.

chasm replied:
I agree. If it's just being passed down from onClick to onClick, then call it onClick. If you get an onClick function in the props and then you want to wrap it in another function that, for example, decides when to call it, then call that wrapper handleClick. I wish I could say that I'm super consistent with this myself, but...

nathanbowang replied:
Weird, I would have thought I rename this prop to handleClick. Thanks for explaining. Definitely, will update in the next commit.

reubenberghan commented on components/Main/NewsletterSignUp/SignUpForm/index.tsx:
```diff
@@ -1,37 +1,22 @@
-      </button>
-    </form>
+    <Form onSubmit={handleSubmit}>
+      <InputField name="email" placeholder="Enter Email" required type="email" />
```
One thing that appears to be missing from our `InputField` and `TextareaField` components APIs is exposing an `onChange` prop to allow these components to be controlled by the form. Right now, without interacting with the DOM directly, we don't have a way to manage the state of these from the `Form` components.

chasm replied:
Agreed. Also, how are we handling ARIA attributes?

natebowang replied:
Weird, I would have thought I rename this prop to handleClick. Thanks for explaining. Definitely, will update in the next commit.

nathanbowang replied:
If there is a label element for the input/textarea component, just label, no aria-label. If no label element, add an aria-label on the input/textarea component.

reubenberghan reviewed (changes requested):
The main thing for me would be to expose an `onChange` prop on the input and textarea fields.

reubenberghan commented on components/Main/ContactUs/index.tsx:
```diff
@@ -1,8 +1,7 @@
-import style from './ContactUs.module.css'
+import SuccessConfirmation from '../form/SuccessConfirmation'
 
 export default function ContactUs(): JSX.Element {
```
Wondering whether this component can be generalised and shared between this and the `NewsletterSignUp`?

nathanbowang replied:
I'm with you the structure of the two pieces of code is similar. However, most attributes are different. I think the structure is clear and can self-explain now. And I'm worried if we generalize the structure to another layer, the attributes will mix together and lost their relationship. What do you think? 
![image](https://user-images.githubusercontent.com/30286194/119241823-71540480-bbad-11eb-8889-4a4fbe6b980f.png)

reubenberghan replied:
No this is pretty much what I meant. If I'm understanding you correctly by attributes you are talking about props? So yeah I would assume things like `title`, `description`, and the children would be different and that the structure is the abstraction. A little unsure of the renaming from `isSuccessful` to `isFilling`, what is the reasoning here?

nathanbowang replied:
Ok, I will abstract the structure in a shared component.

nathanbowang replied:
As to renaming, that's because I want to use true as default, and move `SignUpForm` in front of `SuccessConfirmation`. Which one do you think is better/common?

nathanbowang replied:
Oh, I realize 'isFilling' is not a good name. Since `!isFilling` doesn't mean successfully submitted. I will revert this change.

reubenberghan replied:
For me what makes the most sense is probably `isSubmitted` as it describes what has happened. `isFilling` seems more applicable for while the user is filling out the form and `isSuccessful` would be more like if the response from the back end was successful.

nathanbowang replied:
I haven't done all updates. I'll let you know when done. I realize how forms can be tricky now.

chasm replied:
Beware the gerunds. Words with -ing endings generally mean something that is in progress, while -ed endings mean something that is completed.

nathanbowang committed:
fix: expose setvalue and value prop on the input and textarea component

Expose setValue and value prop on the InputField and TextareaField
component.
Update storybook.
Rename onClick prop.

nathanbowang committed:
refactor(newsletter-sign-up): wrap body, email, and name in an object

nathanbowang committed:
refactor: rename on click handler

nathanbowang committed:
refactor: generalize a step controller component

nathanbowang committed:
refactor(step-switch): refactor the step-switch component

nathanbowang committed:
refactor(step-switch): rename stories' title

nathanbowang committed:
refactor(newsletter-sign-up): make the newsletter sign up component use the generic step switch component

nathanbowang committed:
refactor(contact-us): make the contact us component use the generic step switch component

natebowang commented:
That's all my updates for now. I generalized a component to control the steps.

nathanbowang committed:
refactor: add the finishing touches

Added some finishing touches to the NewsletterSignUp, ContactUs,
StepSwitch component.

natebowang commented:
The last commit, I won't touch this PR from now.

nathanbowang committed:
fix: fix a bug of the mod utility function

natebowang commented:
The last commit. I won't touch this PR from now.

nathanbowang committed:
feat(textarea-field): replace the hard limit with a warning

nathanbowang committed:
refactor: rename handle click prop and add aria-label on input component

nathanbowang committed:
refactor: add aria-label on the text field component

nathanbowang committed:
fix: update textarea length to 500
